src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,162 @@
+
     @Override
     public void add(T value) {
+        if (head == null && tail == null) {
+            head = new Node(null, value, null);
+            tail = head;
+            size++;
+        } else {
+            head.next = new Node<>(head, value, null);
+            head = head.next;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == 0 && head == null && tail == null) {
+            head = new Node(null, value, null);
+            tail = head;
+        } else if (index == 0) {
+            Node<T> nodeToAdd = new Node<>(null, value, tail);
+            tail.prev = nodeToAdd;
+            tail = nodeToAdd;
+        } else if (index == size) {
+            add(value);
+            size--;
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> nodeToAdd = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = nodeToAdd;
+            currentNode.prev = nodeToAdd;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T el : list) {
+            add(el);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> currentNode = getNode(index);
+        Node<T> nodeToSet = new Node<>(currentNode.prev, value, currentNode.next);
+        if (currentNode.next != null) {
+            currentNode.next.prev = nodeToSet;
+        } else {
+            head = nodeToSet;
+        }
+        if (currentNode.prev != null) {
+            currentNode.prev.next = nodeToSet;
+        } else {
+            tail = nodeToSet;
+        }
+        return currentNode.item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> currentNode = getNode(index);
+
+        remove(currentNode.item);
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = getNode(object);
+        if (currentNode == null) {
+            return false;
+        }
+        if (currentNode.next != null) {
+            currentNode.next.prev = currentNode.prev;
+        } else {
+            head = currentNode.prev;
+        }
+        if (currentNode.prev != null) {
+            currentNode.prev.next = currentNode.next;
+        } else {
+            tail = currentNode.next;
```

Comment:
Be careful with linking Nodes. Ensure that you add mutual links from both sides when you insert a new Node, and that you remove all the links when you remove a Node.